File: NyaLauncher.Core/Launch/LaunchInfo.cs
using NyaLauncher.Core.Launch.Auth;
using System.Diagnostics;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace NyaLauncher.Core.Launch;
    /// <summary>
    /// 离线启动所需的外部配置。版本文件、依赖库和资源应已存在于游戏目录。
    /// </summary>
    public sealed class MinecraftLaunchOptions
    {
        public required string MinecraftDirectory { get; init; }

        /// <summary>
        /// 可选的实例游戏目录，用于隔离 mods、config、saves 等内容。
        /// 为空时使用 MinecraftDirectory。
        /// </summary>
        public string? GameDirectory { get; init; }

        public required string VersionId { get; init; }

        /// <summary>
        /// 启动使用的账号；可为离线账号或正版（Microsoft）账号。
        /// </summary>
        public required IMinecraftAccount Account { get; init; }

        /// <summary>
        /// 可选的 Java 可执行文件。为空时依次检查 NYALAUNCHER_JAVA、JAVA_HOME 和 PATH。
        /// </summary>
        public string? JavaExecutable { get; init; }

        /// <summary>
        /// 可选的 Minecraft runtime 根目录；启动器会递归查找并选择版本要求的 Java。
        /// </summary>
        public string? JavaRuntimeDirectory { get; init; }

        public int MinimumMemoryMb { get; init; } = 512;

        public int MaximumMemoryMb { get; init; } = 4096;

        public int WindowWidth { get; init; } = 854;

        public int WindowHeight { get; init; } = 480;

        public string LauncherName { get; init; } = "NyaLauncher";

        public string LauncherVersion { get; init; } = "0.1.0";

        public IReadOnlyList<string> AdditionalJvmArguments { get; init; } = [];

        public IReadOnlyList<string> AdditionalGameArguments { get; init; } = [];

        /// <summary>
        /// 启动过程中的文本日志回调（如 Java 自动下载阶段的进度提示）；可为空。
        /// </summary>
        public Action<string>? LogCallback { get; init; }

        /// <summary>
        /// 返回一个除账号外其余配置完全相同的副本，用于在启动前替换账号。
        /// </summary>
        public MinecraftLaunchOptions WithAccount(IMinecraftAccount account) => new()
        {
            MinecraftDirectory = MinecraftDirectory,
            GameDirectory = GameDirectory,
            VersionId = VersionId,
            Account = account,
            JavaExecutable = JavaExecutable,
            JavaRuntimeDirectory = JavaRuntimeDirectory,
            MinimumMemoryMb = MinimumMemoryMb,
            MaximumMemoryMb = MaximumMemoryMb,
            WindowWidth = WindowWidth,
            WindowHeight = WindowHeight,
            LauncherName = LauncherName,
            LauncherVersion = LauncherVersion,
            AdditionalJvmArguments = AdditionalJvmArguments,
            AdditionalGameArguments = AdditionalGameArguments,
            LogCallback = LogCallback
        };
    }
    /// <summary>
    /// 可供 Minecraft 启动使用的统一账号抽象。
    /// 离线账号与正版（Microsoft）账号都实现此接口。
    /// </summary>
    public interface IMinecraftAccount
    {
        /// <summary>游戏内玩家名（auth_player_name）。</summary>
        string Username { get; }

        /// <summary>
        /// 用于 auth_uuid 的 UUID 字符串。
        /// 正版账号为带连字符的档案 UUID，离线账号为无连字符的离线 UUID。
        /// </summary>
        string Uuid { get; }

        /// <summary>用于 auth_access_token 的访问令牌；离线账号固定为 "0"。</summary>
        string AccessToken { get; }

        /// <summary>用于 user_type 的启动参数：离线为 "legacy"，正版为 "msa"。</summary>
        string UserType { get; }

        /// <summary>用于 auth_xuid 的 Xbox 用户 ID；离线账号为空字符串。</summary>
        string XboxUserId { get; }
    }
    /// <summary>
    /// 不包含任何访问令牌的离线 Minecraft 账号。
    /// </summary>
    public sealed record OfflineAccount : IMinecraftAccount
    {
        private static readonly Regex UsernamePattern =
            new("^[A-Za-z0-9_]{1,16}$", RegexOptions.CultureInvariant);

        private OfflineAccount(string username, string uuid)
        {
            Username = username;
            Uuid = uuid;
        }

        public string Username { get; }

        /// <summary>
        /// 与服务端离线模式一致的 32 位无连字符 UUID。
        /// </summary>
        public string Uuid { get; }

        public string AccessToken => "0";

        public string UserType => "legacy";

        public string XboxUserId => string.Empty;

        public static OfflineAccount Create(string username)
        {
            var normalized = username.Trim();
            if (!UsernamePattern.IsMatch(normalized))
            {
                throw new ArgumentException(
                    "离线用户名必须为 1–16 位，只能包含英文字母、数字和下划线。",
                    nameof(username));
            }

            var source = Encoding.UTF8.GetBytes($"OfflinePlayer:{normalized}");
            var hash = MD5.HashData(source);

            // Java UUID.nameUUIDFromBytes 使用 UUID v3，并设置 RFC 4122 variant。
            hash[6] = (byte)((hash[6] & 0x0f) | 0x30);
            hash[8] = (byte)((hash[8] & 0x3f) | 0x80);

            return new OfflineAccount(normalized, Convert.ToHexStringLower(hash));
        }
    }

    /// <summary>
    /// Minecraft启动约定
    /// </summary>
    public interface IOfflineMinecraftLauncher
    {
        Task<MinecraftLaunchResult> LaunchAsync(
            MinecraftLaunchOptions options,
            CancellationToken cancellationToken = default);
    }
    /// <summary>
    /// 使用正版（Microsoft）账号启动 Minecraft 的入口。
    /// 登录与令牌刷新由 <see cref="Auth.IMicrosoftAuthenticator"/> 完成，
    /// 本类负责校验令牌并复用现有的离线启动管线（进程构造、参数构建、资源解析）。
    /// </summary>
    public interface IMicrosoftMinecraftLauncher
    {
        Task<MinecraftLaunchResult> LaunchAsync(
            MicrosoftAccount account,
            MinecraftLaunchOptions options,
            CancellationToken cancellationToken = default);
    }
    /// <summary>
    /// Minecraft启动错误时抛出的异常类
    /// </summary>
    public sealed class MinecraftLaunchException : Exception
    {
        public MinecraftLaunchException(string message)
            : base(message)
        {
        }

        public MinecraftLaunchException(string message, Exception innerException)
            : base(message, innerException)
        {
        }
    }
    /// <summary>
    /// Minecraft启动计划
    /// 启动Minecraft时，需要传递的参数
    /// </summary>
    /// <param name="JavaExecutable">Java可执行文件路径</param>
    /// <param name="WorkingDirectory">工作目录(.minecraft文件夹)</param>
    /// <param name="NativeDirectory">库目录</param>
    /// <param name="RequiredJavaMajorVersion">所需的Java主版本</param>
    /// <param name="Arguments">启动参数</param>
    internal sealed record MinecraftLaunchPlan(
    string JavaExecutable,
    string WorkingDirectory,
    string NativeDirectory,
    int? RequiredJavaMajorVersion,
    IReadOnlyList<string> Arguments);
    /// <summary>
    /// 
    /// </summary>
    /// <param name="Process"></param>
    /// <param name="VersionId"></param>
    /// <param name="Username"></param>
    /// <param name="RequiredJavaMajorVersion"></param>

    /// <summary>
    /// Minecraft实例启动后结果
    /// </summary>
    public sealed record MinecraftLaunchResult(
    Process Process,
    string VersionId,
    string Username,
    int? RequiredJavaMajorVersion);

File: internal/launch/launch_info.go
package launch

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os/exec"
	"regexp"
	"strings"

	"nyalauncher/internal/auth"
)

// MinecraftAccount 可供 Minecraft 启动使用的统一账号抽象。
// C# 中接口定义在 Launch 命名空间；Go 版实际定义在 auth 包（避免循环依赖），此处别名。
type MinecraftAccount = auth.MinecraftAccount

// offlineUsernamePattern 离线用户名：1–16 位，只能包含英文字母、数字和下划线。
var offlineUsernamePattern = regexp.MustCompile(`^[A-Za-z0-9_]{1,16}$`)

// OfflineAccount 不包含任何访问令牌的离线 Minecraft 账号。
type OfflineAccount struct {
	username string
	// uuid 与服务端离线模式一致的 32 位无连字符 UUID。
	uuid string
}

// NewOfflineAccount 构造离线账号；用户名非法时返回错误。
func NewOfflineAccount(username string) (*OfflineAccount, error) {
	normalized := strings.TrimSpace(username)
	if !offlineUsernamePattern.MatchString(normalized) {
		return nil, fmt.Errorf("离线用户名必须为 1–16 位，只能包含英文字母、数字和下划线。（输入：%q）", username)
	}
	return &OfflineAccount{
		username: normalized,
		uuid:     offlineUuid(normalized),
	}, nil
}

// MustOfflineAccount 构造离线账号；用户名非法时 panic（仅用于内部已校验路径）。
func MustOfflineAccount(username string) *OfflineAccount {
	account, err := NewOfflineAccount(username)
	if err != nil {
		panic(err)
	}
	return account
}

func (a *OfflineAccount) AccountKind() string        { return "offline" }
func (a *OfflineAccount) AccountUsername() string    { return a.username }
func (a *OfflineAccount) AccountUuid() string        { return a.uuid }
func (a *OfflineAccount) AccountAccessToken() string { return "0" }
func (a *OfflineAccount) AccountUserType() string    { return "legacy" }
func (a *OfflineAccount) AccountXboxUserId() string  { return "" }

// offlineUuid 计算 OfflinePlayer:{name} 的 UUID v3（Java UUID.nameUUIDFromBytes
// 语义：MD5 + 设置 RFC 4122 version/variant 位），输出 32 位无连字符小写。
func offlineUuid(username string) string {
	hash := md5.Sum([]byte("OfflinePlayer:" + username))
	bytes := hash[:]
	bytes[6] = (bytes[6] & 0x0f) | 0x30
	bytes[8] = (bytes[8] & 0x3f) | 0x80
	return hex.EncodeToString(bytes)
}

// MinecraftLaunchOptions 离线启动所需的外部配置。版本文件、依赖库和资源应已存在于游戏目录。
type MinecraftLaunchOptions struct {
	// MinecraftDirectory Minecraft 根目录（.minecraft）。
	MinecraftDirectory string
	// GameDirectory 可选的实例游戏目录，用于隔离 mods、config、saves 等内容。
	// 为空时使用 MinecraftDirectory。
	GameDirectory string
	// VersionId 要启动的版本 ID。
	VersionId string
	// Account 启动使用的账号；可为离线账号或正版（Microsoft）账号。
	Account MinecraftAccount
	// JavaExecutable 可选的 Java 可执行文件。为空时依次检查 NYALAUNCHER_JAVA、
	// JAVA_HOME 和 PATH。
	JavaExecutable string
	// JavaRuntimeDirectory 可选的 Minecraft runtime 根目录；启动器会递归查找并
	// 选择版本要求的 Java。
	JavaRuntimeDirectory    string
	MinimumMemoryMb         int
	MaximumMemoryMb         int
	WindowWidth             int
	WindowHeight            int
	LauncherName            string
	LauncherVersion         string
	AdditionalJvmArguments  []string
	AdditionalGameArguments []string
	// Transform 已启用插件贡献的启动变换（Java 路径覆盖、类路径与参数前后插入、
	// 环境变量等）。nil 等价于空变换，即不改变原有启动行为。
	Transform *MinecraftLaunchTransform
	// LogCallback 启动过程中的文本日志回调（如 Java 自动下载阶段的进度提示）；可为空。
	LogCallback func(string)
	// GameOutputCallback 游戏进程 stdout/stderr 行回调（Go 移植新增：C# 由
	// GameLaunchService 直接订阅 Process 事件，Go 版进程由本包创建，故以回调注入）。
	// 第一个参数为行文本，第二个参数为是否 stderr。可为空。
	GameOutputCallback func(line string, isStderr bool)
}

// WithAccount 返回一个除账号外其余配置完全相同的副本，用于在启动前替换账号。
func (o MinecraftLaunchOptions) WithAccount(account MinecraftAccount) MinecraftLaunchOptions {
	copied := o
	copied.Account = account
	return copied
}

// MinecraftLaunchPlan Minecraft 启动计划：进程拉起所需的全部参数。
type MinecraftLaunchPlan struct {
	JavaExecutable           string
	WorkingDirectory         string
	NativeDirectory          string
	RequiredJavaMajorVersion *int
	Arguments                []string
	// EnvironmentVariables 注入子进程的环境变量（对应 C# Dictionary<string,string?> 的非 null 项）。
	EnvironmentVariables map[string]string
	// RemovedEnvironmentVariables 需要从子进程环境中移除的变量名
	//（对应 C# 字典中值为 null 的项）。
	RemovedEnvironmentVariables map[string]bool
}

// MinecraftLaunchResult Minecraft 实例启动后的结果。
type MinecraftLaunchResult struct {
	// Cmd 已启动的 Java 进程；调用方不得重复调用 Wait（本包在启动时已启动
	// 退出观察 goroutine，退出状态经 Exit 通道发布）。
	Cmd *exec.Cmd
	// exitCh 进程退出后收到一次退出结果（Wait 的返回值），容量 1。
	exitCh                   chan error
	VersionId                string
	Username                 string
	RequiredJavaMajorVersion *int
}

// Pid 游戏进程 ID；进程未就绪时返回 0。
func (r *MinecraftLaunchResult) Pid() int {
	if r == nil || r.Cmd == nil || r.Cmd.Process == nil {
		return 0
	}
	return r.Cmd.Process.Pid
}

// Exit 进程退出通道：进程结束后收到一个退出错误（正常退出为 nil）。
func (r *MinecraftLaunchResult) Exit() <-chan error {
	if r.exitCh == nil {
		ch := make(chan error, 1)
		ch <- nil
		return ch
	}
	return r.exitCh
}

// Kill 结束游戏进程（不包含进程树）。
func (r *MinecraftLaunchResult) Kill() error {
	if r == nil || r.Cmd == nil || r.Cmd.Process == nil {
		return fmt.Errorf("进程未在运行")
	}
	return r.Cmd.Process.Kill()
}

// IOfflineMinecraftLauncher 离线启动器约定。
type IOfflineMinecraftLauncher interface {
	Launch(ctx context.Context, options MinecraftLaunchOptions) (*MinecraftLaunchResult, error)
}

// IMicrosoftMinecraftLauncher 使用正版（Microsoft）账号启动 Minecraft 的入口。
// 登录与令牌刷新由 auth 包的认证器完成，实现类负责校验令牌并复用现有的
// 离线启动管线（进程构造、参数构建、资源解析）。
type IMicrosoftMinecraftLauncher interface {
	Launch(ctx context.Context, account *auth.MicrosoftAccount, options MinecraftLaunchOptions) (*MinecraftLaunchResult, error)
}
